docs/notebooks/inequality_constraints.pct.py

```diff
@@ -146,6 +148,94 @@ def create_bo_model(data):
 )
 plt.show()
 
+
+# %% [markdown]
+# ## Define a batch acquisition rule
+#
+# It is sometimes beneficial to query several points at a time instead of one. We show here how to create ad-hoc extensions of the previous acquistion function and rule.
+
+# %%
+class BatchExpectedConstrainedImprovement(trieste.acquisition.BatchAcquisitionFunctionBuilder):
+    def __init__(
+            self,
+            sample_size,
+            threshold,
+    ):
+        self._sample_size = sample_size
+        self.threshold = threshold
+
+    def prepare_acquisition_function(self, datasets, models):
+        objective_model = models[OBJECTIVE]
+        objective_dataset = datasets[OBJECTIVE]
+
+        samplers = {tag: trieste.acquisition.BatchReparametrizationSampler(self._sample_size, model)
+                    for tag, model in models.items()}
+
+        pf = trieste.acquisition.probability_of_feasibility(models[CONSTRAINT],
+                                                                     self.threshold,
+                                                                     objective_dataset.query_points)
+        is_feasible = pf >= 0.5
+
+        mean, _ = objective_model.predict(objective_dataset.query_points)
+        eta = tf.reduce_min(tf.boolean_mask(mean, is_feasible), axis=0)
+
+        def batch_efi(at):
+            samples = {tag: tf.squeeze(sampler.sample(at), -1) for tag, sampler in samplers.items()}
+            feasible_mask = samples[CONSTRAINT] < self.threshold  # [N, S, B]
+            improvement = tf.where(feasible_mask, tf.maximum(eta - samples[OBJECTIVE], 0.), 0.)  # [N, S, B]
+            batch_improvement = tf.reduce_max(improvement, axis=-1)  # [N, S]
+            return tf.reduce_mean(batch_improvement, axis=-1, keepdims=True)  # [N, 1]
+
+        return batch_efi
+
+
+num_query_points = 4
+batch_eci = BatchExpectedConstrainedImprovement(50, Sim.threshold)
+batch_rule = trieste.acquisition.rule.BatchAcquisitionRule(num_query_points, batch_eci)
+
+# %% [markdown]
+# We can now run the BO loop as before; note that we here also query 20 points, but in 5 batches of 4 points.
+
+initial_models = {
+    OBJECTIVE: create_bo_model(initial_data[OBJECTIVE]),
+    CONSTRAINT: create_bo_model(initial_data[CONSTRAINT]),
+}
+
+num_steps = 5
+batch_data = bo.optimize(
+    num_steps, initial_data, initial_models, batch_rule, track_state=False
+).try_get_final_datasets()
+
+# %% [markdown]
+# We visualise the resulting data as before.
+batch_constraint_data = batch_data[CONSTRAINT]
+new_batch_query_points = batch_constraint_data.query_points[-num_query_points * num_steps:]
+new_batch_observations = batch_constraint_data.observations[-num_query_points * num_steps:]
+new_batch_data = (new_batch_query_points, new_batch_observations)
```

Comment:
not something for this PR, but I wonder if we should support indexing on `Dataset`. This would then be
```
new_batch_data = batch_constraint_data[-num_query_points * num_steps:].astuple()
```